Sources/Bot/Services/DispatchService.swift

```diff
@@ -0,0 +1,208 @@
+        for mergeServiceForBranch in mergeServices.values {
+            mergeServiceForBranch.statusChecksCompletionObserver.send(value: event)
+        }
+    }
+
+    @discardableResult
+    private func makeMergeService(targetBranch: String, scheduler: DateScheduler, initialPullRequests: [PullRequest] = []) -> MergeService {
+        let mergeService = MergeService(
+            targetBranch: targetBranch,
+            integrationLabel: integrationLabel,
+            topPriorityLabels: topPriorityLabels,
+            requiresAllStatusChecks: requiresAllStatusChecks,
+            statusChecksTimeout: statusChecksTimeout,
+            initialPullRequests: initialPullRequests,
+            logger: logger,
+            gitHubAPI: gitHubAPI,
+            scheduler: scheduler
+        )
+        self.mergeServices[targetBranch] = mergeService
+        self.healthcheck.startMonitoring(mergeServiceHealthcheck: mergeService.healthcheck)
+        mergeServiceLifecycleObserver?.send(value: .created(mergeService))
+        mergeService.state.producer
+            .observe(on: scheduler)
+            .startWithValues { [weak self, service = mergeService] state in
+                self?.mergeServiceLifecycleObserver?.send(value: .stateChanged(service))
+                if state.status == .idle {
+                    self?.mergeServices[targetBranch] = nil
+                    self?.mergeServiceLifecycleObserver?.send(value: .destroyed(service))
+                }
+            }
+
+        return mergeService
+    }
+}
+
+extension DispatchService {
+    public enum MergeServiceLifecycleEvent {
+        case created(MergeService)
+        case destroyed(MergeService)
+        case stateChanged(MergeService)
+    }
+}
+
+extension DispatchService {
+    public var queuesDescription: String {
+        guard !mergeServices.isEmpty else {
+            return "No PR pending, all queues empty."
+        }
+        return mergeServices.map { (entry: (key: String, value: MergeService)) -> String in
+            """
+            ## Merge Queue for target branch: \(entry.key) ##
+
+            \(entry.value.state.value)
+            """
+        }.joined(separator: "\n\n")
+    }
+
+    public var queueStates: [MergeService.State] {
+        return self.mergeServices.values
+            .map { $0.state.value }
+            .sorted { (lhs, rhs) in
+                lhs.targetBranch < rhs.targetBranch
+        }
+    }
+}
+
+// MARK: - Healthcheck
+
+extension DispatchService {
+    public final class Healthcheck {
+
+        public enum Reason: Error, Equatable {
+            case potentialDeadlock
+        }
+
+        public enum Status: Equatable {
+            case ok
+            case unhealthy(Reason)
+        }
+
+        public var status: Property<Status> { return Property(_status) }
+
+        private let scheduler: DateScheduler
+        private var producers: [String: SignalProducer<MergeService.Healthcheck.Status, NoError>] = [:]
```

Comment:
This also needs to be thread-safe. However, considering the current usage of this root health check, and the underlying implementation which is a bit complex, I think we can live with something more simple like having a function `func currentStatus() -> Status` that internally uses this dictionary and queries each merge service's healthcheck and `reduce`s everything into one consolidated status, what do you think @AliSoftware ?

Reply:
Agreed, I was actually not very happy with this implementation and this kinda dirty trick of disposing and recreating the producer, so I'm happy with an on-demand impl instead 👍

Reply:
Fixed in 80e3d6300f654dc5e2881d013c48191cd2542bde
This indeed simplifies the code _a lot_ 👍